File: src/providers/deepinfra.js
/**
 * DeepInfra provider for speech-to-text transcription.
 */

import fs from 'fs';
import path from 'path';

export class DeepInfraProvider {
  constructor({ apiKey, model = 'openai/whisper-large-v3' }) {
    this.name = 'deepinfra';
    this.apiKey = apiKey;
    this.model = model;
  }

  async transcribe({ audioPath }) {
    const audioData = fs.readFileSync(audioPath);
    const filename = path.basename(audioPath);

    const ext = path.extname(audioPath).toLowerCase();
    const mimeTypes = {
      '.mp3': 'audio/mpeg',
      '.wav': 'audio/wav',
      '.m4a': 'audio/m4a',
      '.ogg': 'audio/ogg',
      '.flac': 'audio/flac',
      '.webm': 'audio/webm',
      '.mp4': 'audio/mp4',
      '.mpeg': 'audio/mpeg',
      '.mpga': 'audio/mpeg',
    };
    const contentType = mimeTypes[ext] || 'audio/mpeg';

    const formData = new FormData();
    const blob = new Blob([audioData], { type: contentType });
    formData.append('audio', blob, filename);

    const response = await fetch(
      `https://api.deepinfra.com/v1/inference/${this.model}`,
      {
        method: 'POST',
        headers: {
          Authorization: `Bearer ${this.apiKey}`,
        },
        body: formData,
      }
    );

    if (!response.ok) {
      const errorText = await response.text();
      throw new Error(
        `DeepInfra API error: ${response.status} - ${errorText}`
      );
    }

    const result = await response.json();
    return {
      text: result.text?.trim() || '',
      provider: this.name,
    };
  }
}

File: python/speech_to_text/providers/deepinfra.py
"""DeepInfra provider for transcription."""

from pathlib import Path
from typing import Any, Dict

from .base import BaseProvider


class DeepInfraProvider(BaseProvider):
    """DeepInfra transcription provider."""

    def __init__(self, options: Dict[str, Any]):
        """Initialize DeepInfra provider.

        Args:
            options: Configuration options
        """
        super().__init__(options)
        self.name = "deepinfra"
        self.model = options.get("model", "openai/whisper-large-v3")

    async def transcribe(self, options: Dict[str, Any]) -> Dict[str, Any]:
        """Transcribe audio using DeepInfra API.

        Args:
            options: Dictionary containing audio_path

        Returns:
            Transcription result dictionary
        """
        import aiohttp

        audio_path = Path(options["audio_path"])
        filename = audio_path.name

        # Determine MIME type
        ext = audio_path.suffix.lower()
        mime_types = {
            ".mp3": "audio/mpeg",
            ".wav": "audio/wav",
            ".m4a": "audio/m4a",
            ".ogg": "audio/ogg",
            ".flac": "audio/flac",
            ".webm": "audio/webm",
            ".mp4": "audio/mp4",
        }
        content_type = mime_types.get(ext, "audio/mpeg")

        # Create multipart form data
        data = aiohttp.FormData()
        data.add_field(
            "audio",
            open(audio_path, "rb"),
            filename=filename,
            content_type=content_type,
        )

        headers = {"Authorization": f"Bearer {self.api_key}"}

        async with aiohttp.ClientSession() as session:
            async with session.post(
                f"https://api.deepinfra.com/v1/inference/{self.model}",
                headers=headers,
                data=data,
            ) as response:
                if not response.ok:
                    error_text = await response.text()
                    raise Exception(
                        f"DeepInfra API error: {response.status} - {error_text}"
                    )

                result = await response.json()
                return {
                    "text": result.get("text", "").strip(),
                    "provider": self.name,
                }
